spec/compiler/parser/parser_spec.cr

```diff
@@ -226,11 +226,11 @@ module Crystal
     it_parses "def foo(var = 1); end", Def.new("foo", [Arg.new("var", 1.int32)])
     it_parses "def foo(var : Int); end", Def.new("foo", [Arg.new("var", restriction: "Int".path)])
     it_parses "def foo(var : self); end", Def.new("foo", [Arg.new("var", restriction: Self.new)])
-    it_parses "def foo(var : self?); end", Def.new("foo", [Arg.new("var", restriction: Crystal::Union.new([Self.new, Path.global("Nil")] of ASTNode))])
+    it_parses "def foo(var : self?); end", Def.new("foo", [Arg.new("var", restriction: Crystal::Generic.new(Path.global("Union"), [Self.new, Path.global("Nil")] of ASTNode))])
```

Comment:
Why does this change? Same for other place where we now produce a Generic instead of Union. I think Union is simpler...?

Reply:
`Generic` has `question` property to represent suffix `?`. (see [here](https://github.com/crystal-lang/crystal/blob/2cbe65b0a13c8a2ce91bd4799f5b96b8e1953097/src/compiler/crystal/syntax/ast.cr#L1448-L1449)). So I think it is better than using `Union`.

Reply:
The thing is that we have code in `restrictions.cr` for Union and I think it should work the same as for Generic, but I'm not sure. So changing this might indeed break someones code because we are changing behavior.

Reply:
I'm not sure too. But your concern is semantic analyzer's bug, and this is also parser's bug fix.
For instance, currently `Foo(Bar?)?` is parsed to `Generic.new(Path.global("Union"), [Generic.new(Path.new("Foo"), [Union.new([Path.new("Bar"), Path.global("Nil")])]), Path.global("Nil")])` (inner `Bar?` uses `Union` and outer `Foo(...)?` uses `Generic`.)

When such a bug is found really, we can fix analyzer correctly. Leaving a certain bug seems not good.

Reply:
Okay, let's wait and see if bugs happen then.